src/main/java/org/testng/GuiceHelper.java

```diff
@@ -121,12 +124,29 @@ public static Injector createInjector(ITestContext context,
     Class<? extends IModuleFactory> factory = guice.moduleFactory();
     if (factory != IModuleFactory.class) {
       IModuleFactory factoryInstance = parentInjector.getInstance(factory);
-      Module moduleClass = factoryInstance.createModule(context, testClass);
-      if (moduleClass != null) {
-        result.add(moduleClass);
+      Module module = factoryInstance.createModule(context, testClass);
+      if (module != null) {
+        result.add(module);
       }
     }
-
+    result.addAll(getSpiModules());
     return result;
   }
-}
+
+  private List<Module> getSpiModules() {
+    List<Module> spiModules = new ArrayList<>();
+    for (IModule module : ServiceLoader.load(IModule.class)) {
+      Class<? extends IModule> moduleClass = module.getClass();
+      List<Module> cachedModules = context.getGuiceModules(moduleClass);
+      if (cachedModules == null) {
+        // skip because ITestContext does not have a setter method for Guice module list
+      } else if ( cachedModules.isEmpty() ) {
+        cachedModules.add(module);
+        spiModules.add(module);
+      } else {
+        spiModules.add(cachedModules.get(0));
```

Comment:
I didn't understand this part wherein we are resorting to just getting the first element in the list. So what would happen to the rest of them ?
I am not that familiar with Guice as a DI library. So please elaborate this part.

Reply:
I assumed there is always a single instance of Module, but you are right, it should be addAll to be consistent with org.testng.GuiceHelper#getModules.